site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2360,6 +2361,109 @@ public function getAllAnonIds() {
         return $this->course_db->rows();
     }
 
+    /**
+     * Generate notifcation rows
+     *
+     * @param Notification $notification
+     */
+    public function pushNotification($notification){
+        $params = array();
+        $params[] = $notification->getComponent();
+        $params[] = $notification->getNotifyMetadata();
+        $params[] = $notification->getNotifyContent();
+        $params[] = $notification->getNotifySource();
+
+        if(empty($notification->getNotifyTarget())) {
+            // Notify all users
+            $target_users_query = "SELECT user_id FROM users";
+        } else {
+            // To a specific user
+            $params[] = $notification->getNotifyTarget();
+            $target_users_query = "SELECT ?::text as user_id";
+        }
+
+        if($notification->getNotifyNotToSource()){
+            $ignore_self_query = "WHERE user_id <> ?";
+            $params[] = $notification->getNotifySource();
+        }
+        else {
+            $ignore_self_query = "";
+        }
+        $this->course_db->query("INSERT INTO notifications(type, metadata, content, created_at, from_user_id, to_user_id)
+                    SELECT ?, ?, ?, current_timestamp, ?, user_id as to_user_id FROM ({$target_users_query}) as u {$ignore_self_query}",
+                    $params);
+    }
+
+    /**
+     * Returns notifications for a user
+     *
+     * @param string $user_id
+     * @param bool $show_all
+     * @return array(Notification)
+     */
+    public function getUserNotifications($user_id, $show_all){
+        if($show_all){
+            $seen_status_query = "true";
+        } else {
+            $seen_status_query = "seen_at is NULL";
+        }
+        $this->course_db->query("SELECT id, type, metadata, content,
+                (case when seen_at is NULL then false else true end) as seen,
+                (extract(epoch from current_timestamp) - extract(epoch from created_at)) as elapsed_time, created_at
+                FROM notifications WHERE to_user_id = ? and {$seen_status_query} ORDER BY created_at DESC", array($user_id));
+        $rows = $this->course_db->rows();
+        $results = array();
+        foreach ($rows as $row) {
+            $results[] = new Notification($this->core, array(
+                    'view_only' => true,
+                    'id' => $row['id'],
+                    'component' => $row['type'],
```

Comment:
Its confusing that you have a `type` field in the database that gets converted to a `component` field in the model, but that model also takes a `type` parameter that means something different.  If you make `type` consistent with `type` in the database and maybe change `type` as it exists in php now to `action` it would be a lot more clear.